<?xml version="1.0" ?>
<!-- =================================================================================== -->
<!-- |    This document was autogenerated by xacro from panda_arm.urdf.xacro           | -->
<!-- |    EDITING THIS FILE BY HAND IS NOT RECOMMENDED                                 | -->
<!-- =================================================================================== -->
<robot name="panda" xmlns:xacro="http://www.ros.org/wiki/xacro">
  <gazebo reference="world">
    <!--     <static>true</static>
    <turnGravityOff>true</turnGravityOff> -->
  </gazebo>
  <link name="world"/>
  <joint name="panda_joint_world" type="fixed">
    <parent link="world"/>
    <child link="panda_link0"/>
    <origin rpy="0 0 0" xyz="0 0 0"/>
  </joint>
  <gazebo>
    <plugin filename="libgazebo_ros_control.so" name="gazebo_ros_control">
      <robotNamespace>panda</robotNamespace>
      <legacyModeNS>true</legacyModeNS>
    </plugin>
  </gazebo>
  <gazebo reference="panda_link0"/>
  <gazebo reference="panda_link1"/>
  <gazebo reference="panda_link2"/>
  <gazebo reference="panda_link3"/>
  <gazebo reference="panda_link4"/>
  <gazebo reference="panda_link5"/>
  <gazebo reference="panda_link6"/>
  <gazebo reference="panda_link7"/>
  <gazebo reference="panda_link8"/>
  <transmission name="panda_joint1_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="panda_joint1">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
      <!-- <hardwareInterface>PositionJointInterface</hardwareInterface> -->
    </joint>
    <actuator name="panda_joint1_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <transmission name="panda_joint2_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="panda_joint2">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
      <!-- <hardwareInterface>PositionJointInterface</hardwareInterface> -->
    </joint>
    <actuator name="panda_joint2_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <transmission name="panda_joint3_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="panda_joint3">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
      <!-- <hardwareInterface>PositionJointInterface</hardwareInterface> -->
    </joint>
    <actuator name="panda_joint3_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <transmission name="panda_joint4_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="panda_joint4">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
      <!-- <hardwareInterface>PositionJointInterface</hardwareInterface> -->
    </joint>
    <actuator name="panda_joint4_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <transmission name="panda_joint5_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="panda_joint5">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
      <!-- <hardwareInterface>PositionJointInterface</hardwareInterface> -->
    </joint>
    <actuator name="panda_joint5_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <transmission name="panda_joint6_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="panda_joint6">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
      <!-- <hardwareInterface>PositionJointInterface</hardwareInterface> -->
    </joint>
    <actuator name="panda_joint6_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <transmission name="panda_joint7_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="panda_joint7">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
      <!-- <hardwareInterface>PositionJointInterface</hardwareInterface> -->
    </joint>
    <actuator name="panda_joint7_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <link name="panda_link0">
    <inertial>
      <mass value="2.92"/>
      <origin rpy="0 0 0" xyz="-0.025566 -2.88e-05 0.057332"/>
      <inertia ixx="0.00782229414331" ixy="-1.56191622996e-05" ixz="-0.00126005738123" iyy="0.0109027971813" iyz="1.08233858202e-05" izz="0.0102355503949"/>
    </inertial>
    <visual>
      <geometry>
        <mesh filename="package://surgical_sim/models/meshes/dae/franka/link0.dae"/>
      </geometry>
      <material name="">
        <color rgba="1 1 1 1"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://surgical_sim/models/meshes/stl/franka/link0.stl"/>
      </geometry>
    </collision>
  </link>
  <link name="panda_link1">
    <inertial>
      <mass value="2.74"/>
      <origin rpy="0 0 0" xyz="0 -0.0324958 -0.0675818"/>
      <inertia ixx="0.0180416958283" ixy="0.0" ixz="0.0" iyy="0.0159136071891" iyz="0.0046758424612" izz="0.00620690827127"/>
    </inertial>
    <visual>
      <geometry>
        <mesh filename="package://surgical_sim/models/meshes/dae/franka/link1.dae"/>
      </geometry>
      <material name="">
        <color rgba="1 1 1 1"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://surgical_sim/models/meshes/stl/franka/link1.stl"/>
      </geometry>
    </collision>
  </link>
  <joint name="panda_joint1" type="revolute">
    <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-2.8973" soft_upper_limit="2.8973"/>
    <origin rpy="0 0 0" xyz="0 0 0.333"/>
    <parent link="panda_link0"/>
    <child link="panda_link1"/>
    <axis xyz="0 0 1"/>
    <limit effort="87" lower="-2.8973" upper="2.8973" velocity="2.1750"/>
  </joint>
  <link name="panda_link2">
    <inertial>
      <mass value="2.74"/>
      <origin rpy="0 0 0" xyz="0 -0.06861 0.0322285"/>
      <inertia ixx="0.0182856182281" ixy="0.0" ixz="0.0" iyy="0.00621358421175" iyz="-0.00472844221905" izz="0.0161514346309"/>
    </inertial>
    <visual>
      <geometry>
        <mesh filename="package://surgical_sim/models/meshes/dae/franka/link2.dae"/>
      </geometry>
      <material name="">
        <color rgba="1 1 1 1"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://surgical_sim/models/meshes/stl/franka/link2.stl"/>
      </geometry>
    </collision>
  </link>
  <joint name="panda_joint2" type="revolute">
    <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-1.7628" soft_upper_limit="1.7628"/>
    <origin rpy="-1.57079632679 0 0" xyz="0 0 0"/>
    <parent link="panda_link1"/>
    <child link="panda_link2"/>
    <axis xyz="0 0 1"/>
    <limit effort="87" lower="-1.7628" upper="1.7628" velocity="2.1750"/>
  </joint>
  <link name="panda_link3">
    <inertial>
      <mass value="2.38"/>
      <origin rpy="0 0 0" xyz="0.0469893 0.0316374 -0.031704"/>
      <inertia ixx="0.00771376630908" ixy="-0.00248490625138" ixz="-0.00332147581033" iyy="0.00989108008727" iyz="-0.00217796151484" izz="0.00811723558464"/>
    </inertial>
    <visual>
      <geometry>
        <mesh filename="package://surgical_sim/models/meshes/dae/franka/link3.dae"/>
      </geometry>
      <material name="">
        <color rgba="1 1 1 1"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://surgical_sim/models/meshes/stl/franka/link3.stl"/>
      </geometry>
    </collision>
  </link>
  <joint name="panda_joint3" type="revolute">
    <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-2.8973" soft_upper_limit="2.8973"/>
    <origin rpy="1.57079632679 0 0" xyz="0 -0.316 0"/>
    <parent link="panda_link2"/>
    <child link="panda_link3"/>
    <axis xyz="0 0 1"/>
    <limit effort="87" lower="-2.8973" upper="2.8973" velocity="2.1750"/>
  </joint>
  <link name="panda_link4">
    <inertial>
      <mass value="2.38"/>
      <origin rpy="0 0 0" xyz="-0.0360446 0.0336853 0.031882"/>
      <inertia ixx="0.00799663881132" ixy="0.00347095570217" ixz="-0.00241222942995" iyy="0.00825390705278" iyz="0.00235774044121" izz="0.0102515004345"/>
    </inertial>
    <visual>
      <geometry>
        <mesh filename="package://surgical_sim/models/meshes/dae/franka/link4.dae"/>
      </geometry>
      <material name="">
        <color rgba="1 1 1 1"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://surgical_sim/models/meshes/stl/franka/link4.stl"/>
      </geometry>
    </collision>
  </link>
  <joint name="panda_joint4" type="revolute">
    <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-3.0718" soft_upper_limit="0.0698"/>
    <!-- soft_upper_limitr should be -0.0698, but that brakes simulation when initial position for this joint is 0.0 -->
    <origin rpy="1.57079632679 0 0" xyz="0.0825 0 0"/>
    <parent link="panda_link3"/>
    <child link="panda_link4"/>
    <axis xyz="0 0 1"/>
    <limit effort="87" lower="-3.0718" upper="0.0698" velocity="2.1750"/>
    <!-- upper should be -0.0698, but that brakes simulation when initial position for this joint is 0.0 -->
  </joint>
  <link name="panda_link5">
    <inertial>
      <mass value="2.74"/>
      <origin rpy="0 0 0" xyz="0 0.0610427 -0.104176"/>
      <inertia ixx="0.030371374513" ixy="6.50283587108e-07" ixz="-1.05129179916e-05" iyy="0.0288752887402" iyz="-0.00775653445787" izz="0.00444134056164"/>
    </inertial>
    <visual>
      <geometry>
        <mesh filename="package://surgical_sim/models/meshes/dae/franka/link5.dae"/>
      </geometry>
      <material name="">
        <color rgba="1 1 1 1"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://surgical_sim/models/meshes/stl/franka/link5.stl"/>
      </geometry>
    </collision>
  </link>
  <joint name="panda_joint5" type="revolute">
    <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-2.8973" soft_upper_limit="2.8973"/>
    <origin rpy="-1.57079632679 0 0" xyz="-0.0825 0.384 0"/>
    <parent link="panda_link4"/>
    <child link="panda_link5"/>
    <axis xyz="0 0 1"/>
    <limit effort="12" lower="-2.8973" upper="2.8973" velocity="2.6100"/>
  </joint>
  <link name="panda_link6">
    <inertial>
      <mass value="1.55"/>
      <origin rpy="0 0 0" xyz="0.0510509 0.009108 0.0106343"/>
      <inertia ixx="0.00303336450376" ixy="-0.000437276865508" ixz="0.000629257294877" iyy="0.00404479911567" iyz="0.000130472021025" izz="0.00558234286039"/>
    </inertial>
    <visual>
      <geometry>
        <mesh filename="package://surgical_sim/models/meshes/dae/franka/link6.dae"/>
      </geometry>
      <material name="">
        <color rgba="1 1 1 1"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://surgical_sim/models/meshes/stl/franka/link6.stl"/>
      </geometry>
    </collision>
  </link>
  <joint name="panda_joint6" type="revolute">
    <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-0.0175" soft_upper_limit="3.7525"/>
    <origin rpy="1.57079632679 0 0" xyz="0 0 0"/>
    <parent link="panda_link5"/>
    <child link="panda_link6"/>
    <axis xyz="0 0 1"/>
    <limit effort="12" lower="-0.0175" upper="3.7525" velocity="2.6100"/>
  </joint>
  <link name="panda_link7">
    <inertial>
      <mass value="0.54"/>
      <origin rpy="0 0 0" xyz="0.0109695 0.0107965 0.0650411"/>
      <inertia ixx="0.000888868887021" ixy="-0.00012239074652" ixz="3.98699829666e-05" iyy="0.000888001373233" iyz="-9.33825115206e-05" izz="0.0007176834609"/>
    </inertial>
    <visual>
      <geometry>
        <mesh filename="package://surgical_sim/models/meshes/dae/franka/link7.dae"/>
      </geometry>
      <material name="">
        <color rgba="1 1 1 1"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://surgical_sim/models/meshes/stl/franka/link7.stl"/>
      </geometry>
    </collision>
  </link>
  <joint name="panda_joint7" type="revolute">
    <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-2.8973" soft_upper_limit="2.8973"/>
    <origin rpy="1.57079632679 0 0" xyz="0.088 0 0"/>
    <parent link="panda_link6"/>
    <child link="panda_link7"/>
    <axis xyz="0 0 1"/>
    <limit effort="12" lower="-2.8973" upper="2.8973" velocity="2.6100"/>
  </joint>
  <link name="panda_link8"/>
  <joint name="panda_joint8" type="fixed">
    <origin rpy="0 0 0" xyz="0 0 0.107"/>
    <parent link="panda_link7"/>
    <child link="panda_link8"/>
    <axis xyz="0 0 0"/>
  </joint>
  <link name="panda_tool_base_link">
    <inertial>
      <origin rpy="0.0 0.0 0.0" xyz="0.0 0.0 0.0"/>
      <mass value="0.001"/>
      <inertia ixx="0.001" ixy="0.0" ixz="0.0" iyy="0.0001" iyz="0.0" izz="0.0001"/>
    </inertial>
  </link>
  <!-- link 3: tool_main_link -->
  <link name="panda_tool_main_link">
    <visual>
      <origin rpy="0 0 1.5708" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://surgical_sim/models/meshes/dae/tool/tool_main_.dae"/>
      </geometry>
    </visual>
    <!-- <collision>
        <origin rpy="0 0 1.5708" xyz="0 0 0"/>
        <geometry>
          <mesh filename="package://surgical_sim/models/meshes/dae/tool/tool_main_.dae"/>      
        </geometry>
      </collision> -->
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="0.1"/>
      <inertia ixx="0.00133395833333" ixy="0" ixz="0" iyy="0.00133395833333" iyz="0" izz="1.25e-06"/>
    </inertial>
  </link>
  <!-- link 4: tool_wrist_link -->
  <link name="panda_tool_wrist_link">
    <visual>
      <origin rpy="0 0 1.5708" xyz="0 0 -0.0091"/>
      <geometry>
        <mesh filename="package://surgical_sim/models/meshes/dae/tool/tool_wrist_link.dae"/>
      </geometry>
    </visual>
    <collision>
      <origin rpy="0 0 1.5708" xyz="0 0 -0.0091"/>
      <geometry>
        <mesh filename="package://surgical_sim/models/meshes/dae/tool/tool_wrist_link.dae"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="0.05"/>
      <inertia ixx="0.01" ixy="0.0" ixz="0.0" iyy="0.01" iyz="0.0" izz="0.005"/>
    </inertial>
  </link>
  <!-- link 4-1: tool_wrist_shaft_link -->
  <link name="panda_tool_wrist_shaft_link">
    <visual>
      <origin rpy="1.5708 0 0" xyz="0 0.00401 0"/>
      <geometry>
        <mesh filename="package://surgical_sim/models/meshes/dae/tool/tool_wrist_shaft_link.dae"/>
      </geometry>
    </visual>
    <collision>
      <origin rpy="1.5708 0 0" xyz="0 0.00401 0"/>
      <geometry>
        <mesh filename="package://surgical_sim/models/meshes/dae/tool/tool_wrist_shaft_link.dae"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="0.05"/>
      <inertia ixx="0.01" ixy="0.0" ixz="0.0" iyy="0.01" iyz="0.0" izz="0.005"/>
    </inertial>
  </link>
  <gazebo reference="panda_tool_main_link">
    <!-- <turnGravityOff>true</turnGravityOff> -->
    <mu1>0.112</mu1>
    <mu2>0.112</mu2>
    <kp>1000000</kp>
    <kd>1</kd>
    <selfCollide>true</selfCollide>
    <maxVel>0.01</maxVel>
    <minDepth>0.003</minDepth>
  </gazebo>
  <gazebo reference="panda_tool_wrist_link">
    <!-- <turnGravityOff>true</turnGravityOff> -->
    <mu1>0.112</mu1>
    <mu2>0.112</mu2>
    <kp>1000000</kp>
    <kd>1</kd>
    <selfCollide>true</selfCollide>
    <maxVel>0.01</maxVel>
    <minDepth>0.003</minDepth>
  </gazebo>
  <gazebo reference="panda_tool_wrist_shaft_link">
    <!-- <turnGravityOff>true</turnGravityOff> -->
    <mu1>0.112</mu1>
    <mu2>0.112</mu2>
    <kp>1000000</kp>
    <kd>1</kd>
    <selfCollide>true</selfCollide>
    <maxVel>0.01</maxVel>
    <minDepth>0.003</minDepth>
  </gazebo>
  <joint name="panda_tool_main_fixed" type="fixed">
    <parent link="panda_tool_base_link"/>
    <child link="panda_tool_main_link"/>
    <axis xyz="0 0 1"/>
    <origin rpy="0 0 0" xyz="0 0 0.0"/>
  </joint>
  <joint name="panda_tool_wrist_fixed" type="fixed">
    <parent link="panda_tool_main_link"/>
    <child link="panda_tool_wrist_link"/>
    <axis xyz="0 0 1"/>
    <origin rpy="0 0 0" xyz="0 0 0.3752"/>
  </joint>
  <!-- joint 4-1: tool roll shaft joint FIXED-->
  <joint name="panda_tool_wrist_shaft_fixed" type="fixed">
    <parent link="panda_tool_wrist_link"/>
    <child link="panda_tool_wrist_shaft_link"/>
    <axis xyz="0 0 1"/>
    <origin rpy="0 0 0" xyz="0 0 0.0"/>
  </joint>
  <gazebo reference="panda_tool_wrist_fixed">
    <!-- <implicitSpringDamper>0</implicitSpringDamper>
			<provideFeedback>1</provideFeedback> -->
  </gazebo>
  <gazebo reference="panda_tool_wrist_shaft_fixed">
    <!-- <implicitSpringDamper>0</implicitSpringDamper>
			<provideFeedback>1</provideFeedback> -->
  </gazebo>
  <!-- link 5: tool_wrist_forceps_link -->
  <link name="panda_tool_wrist_forceps_link">
    <visual>
      <origin rpy="0.00 -1.5708 0.0" xyz="-0.0024 0.00 0.000"/>
      <geometry>
        <mesh filename="package://surgical_sim/models/meshes/stl/tool/tool_wrist_forceps_link_1.stl"/>
      </geometry>
      <material name="gray">
        <color rgba="0.5 0.5 0.9 1.0"/>
      </material>
    </visual>
    <collision>
      <origin rpy="0.00 -1.5708 0.0" xyz="-0.0024 0.00 0.000"/>
      <geometry>
        <mesh filename="package://surgical_sim/models/meshes/stl/tool/tool_wrist_forceps_link_1.stl"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="1"/>
      <inertia ixx="1" ixy="0.0" ixz="0.0" iyy="1" iyz="0.0" izz="1"/>
    </inertial>
  </link>
  <!-- link 6: tool_wrist_forceps_link_shaft -->
  <link name="panda_tool_wrist_forceps_link_shaft">
    <visual>
      <origin rpy="0.00 1.5708 0.0" xyz="0.0 -0.0002 0.000"/>
      <geometry>
        <mesh filename="package://surgical_sim/models/meshes/stl/tool/tool_wrist_forceps_link_1_shaft.stl"/>
      </geometry>
      <material name="gray">
        <color rgba="0.5 0.5 0.9 1.0"/>
      </material>
    </visual>
    <collision>
      <origin rpy="0.00 1.5708 0.0" xyz="0.0 0.00 0.000"/>
      <geometry>
        <mesh filename="package://surgical_sim/models/meshes/stl/tool/tool_wrist_forceps_link_1_shaft.stl"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="1"/>
      <inertia ixx="1" ixy="0.0" ixz="0.0" iyy="1" iyz="0.0" izz="1"/>
    </inertial>
  </link>
  <!-- link 6-1: Outer open angle link 1 -->
  <link name="panda_tool_wrist_forceps_link_2_left">
    <visual>
      <origin rpy="0.0 0 0.0" xyz="0 0 0.00195"/>
      <geometry>
        <mesh filename="package://surgical_sim/models/meshes/stl/tool/tool_wrist_forceps_link_2_left.stl"/>
      </geometry>
    </visual>
    <collision>
      <origin rpy="0.0 0 0.0" xyz="0 0 0.00195"/>
      <geometry>
        <mesh filename="package://surgical_sim/models/meshes/stl/tool/tool_wrist_forceps_link_2_left.stl"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="1"/>
      <inertia ixx="1" ixy="0.0" ixz="0.0" iyy="1" iyz="0.0" izz="1"/>
    </inertial>
  </link>
  <!-- link 6-2: Outer open angle link 2 -->
  <link name="panda_tool_wrist_forceps_link_2_right">
    <visual>
      <origin rpy="0 0 -3.1415926" xyz="0 0 0.00195"/>
      <geometry>
        <mesh filename="package://surgical_sim/models/meshes/stl/tool/tool_wrist_forceps_link_2_right.stl"/>
      </geometry>
    </visual>
    <collision>
      <origin rpy="0 0 -3.1415926" xyz="0 0 0.00195"/>
      <geometry>
        <mesh filename="package://surgical_sim/models/meshes/stl/tool/tool_wrist_forceps_link_2_right.stl"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <mass value="1"/>
      <inertia ixx="1" ixy="0.0" ixz="0.0" iyy="1" iyz="0.0" izz="1"/>
    </inertial>
  </link>
  <link name="panda_tool_tip_link"/>
  <link name="panda_tool_tip_link_ee"/>
  <gazebo reference="panda_tool_wrist_forceps_link">
    <!-- <turnGravityOff>true</turnGravityOff> -->
    <mu1>0.112</mu1>
    <mu2>0.112</mu2>
    <kp>1000000</kp>
    <kd>1</kd>
    <selfCollide>true</selfCollide>
    <maxVel>0.01</maxVel>
    <minDepth>0.003</minDepth>
  </gazebo>
  <gazebo reference="panda_tool_wrist_forceps_link_shaft">
    <!-- <turnGravityOff>true</turnGravityOff> -->
    <mu1>0.112</mu1>
    <mu2>0.112</mu2>
    <kp>1000000</kp>
    <kd>1</kd>
    <selfCollide>true</selfCollide>
    <maxVel>0.01</maxVel>
    <minDepth>0.003</minDepth>
  </gazebo>
  <gazebo reference="panda_tool_wrist_forceps_link_2_left">
    <!-- <turnGravityOff>true</turnGravityOff> -->
    <mu1>0.112</mu1>
    <mu2>0.112</mu2>
    <kp>1000000</kp>
    <kd>1</kd>
    <selfCollide>true</selfCollide>
    <maxVel>0.01</maxVel>
    <minDepth>0.003</minDepth>
  </gazebo>
  <gazebo reference="panda_tool_wrist_forceps_link_2_right">
    <!-- <turnGravityOff>true</turnGravityOff> -->
    <mu1>0.112</mu1>
    <mu2>0.112</mu2>
    <kp>1000000</kp>
    <kd>1</kd>
    <selfCollide>true</selfCollide>
    <maxVel>0.01</maxVel>
    <minDepth>0.003</minDepth>
  </gazebo>
  <gazebo reference="panda_tool_tip_link">
    <!-- <turnGravityOff>true</turnGravityOff> -->
    <mu1>0.112</mu1>
    <mu2>0.112</mu2>
    <kp>1000000</kp>
    <kd>1</kd>
    <selfCollide>true</selfCollide>
    <maxVel>0.01</maxVel>
    <minDepth>0.003</minDepth>
  </gazebo>
  <gazebo reference="panda_tool_tip_link_ee">
    <!-- <turnGravityOff>true</turnGravityOff> -->
    <mu1>0.112</mu1>
    <mu2>0.112</mu2>
    <kp>1000000</kp>
    <kd>1</kd>
    <selfCollide>true</selfCollide>
    <maxVel>0.01</maxVel>
    <minDepth>0.003</minDepth>
  </gazebo>
  <!-- joint 5: tool wrist pitch joint -->
  <joint name="panda_tool_wrist_pitch" type="fixed">
    <parent link="panda_tool_wrist_shaft_link"/>
    <!-- <parent link="${tool_id}_world"/> -->
    <child link="panda_tool_wrist_forceps_link"/>
    <axis xyz="0 0 1"/>
    <origin rpy="-1.5708 -1.5708 3.14159265359" xyz="0 0 0"/>
  </joint>
  <!-- joint 6: Tool wrist yaw joint -->
  <joint name="panda_tool_wrist_yaw" type="fixed">
    <parent link="panda_tool_wrist_forceps_link"/>
    <child link="panda_tool_wrist_forceps_link_shaft"/>
    <axis xyz="0 0 1"/>
    <origin rpy="0.00 -1.5708 -1.5708" xyz="0.0091 0 0"/>
  </joint>
  <!-- joint 6-1: tool open angle 1 -->
  <joint name="panda_tool_wrist_open_angle" type="revolute">
    <parent link="panda_tool_wrist_forceps_link_shaft"/>
    <child link="panda_tool_wrist_forceps_link_2_left"/>
    <axis xyz="0 0 1"/>
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <limit effort="1000" lower="0" upper="0.349065850399" velocity="3.0"/>
  </joint>
  <!-- joint 6-2: tool open angle 2  -->
  <joint name="panda_tool_wrist_open_angle_mimic" type="revolute">
    <parent link="panda_tool_wrist_forceps_link_shaft"/>
    <child link="panda_tool_wrist_forceps_link_2_right"/>
    <axis xyz="0 0 1"/>
    <origin rpy="3.1415926 0 0" xyz="0 0 0"/>
    <limit effort="1000" lower="0" upper="0.349065850399" velocity="3.0"/>
    <mimic joint="panda_tool_wrist_open_angle" multiplier="1"/>
  </joint>
  <!-- tool tip -->
  <joint name="panda_tool_tip_" type="fixed">
    <parent link="panda_tool_wrist_forceps_link_shaft"/>
    <child link="panda_tool_tip_link"/>
    <origin rpy="0 1.57079632679 1.57079632679" xyz="0 0.0202 0"/>
  </joint>
  <joint name="panda_tool_tip_ee" type="fixed">
    <parent link="panda_tool_tip_link"/>
    <child link="panda_tool_tip_link_ee"/>
    <origin rpy="3.14159265359 0 -1.57079632679" xyz="0 0 0"/>
  </joint>
  <transmission name="panda_tool_wrist_open_angle_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="panda_tool_wrist_open_angle">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="panda_tool_wrist_open_angle_motor">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <transmission name="panda_tool_wrist_open_angle_mimic_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="panda_tool_wrist_open_angle_mimic">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="panda_tool_wrist_open_angle_mimic_motor">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <gazebo reference="panda_tool_wrist_pitch">
    <!-- <implicitSpringDamper>0</implicitSpringDamper>
			<provideFeedback>1</provideFeedback> -->
  </gazebo>
  <gazebo reference="panda_tool_wrist_yaw">
    <!-- <implicitSpringDamper>0</implicitSpringDamper>
			<provideFeedback>1</provideFeedback> -->
  </gazebo>
  <gazebo reference="panda_tool_wrist_open_angle">
    <!-- <implicitSpringDamper>0</implicitSpringDamper>
			<provideFeedback>1</provideFeedback> -->
  </gazebo>
  <gazebo reference="panda_tool_wrist_open_angle_mimic">
    <!-- <implicitSpringDamper>0</implicitSpringDamper>
			<provideFeedback>1</provideFeedback> -->
  </gazebo>
  <gazebo reference="panda_tool_tip_">
    <!-- <implicitSpringDamper>0</implicitSpringDamper>
			<provideFeedback>1</provideFeedback> -->
  </gazebo>
  <gazebo reference="panda_tool_tip_ee">
    <!-- <implicitSpringDamper>0</implicitSpringDamper>
			<provideFeedback>1</provideFeedback> -->
  </gazebo>
  <joint name="panda_base_fixed" type="fixed">
    <parent link="panda_link8"/>
    <child link="panda_tool_base_link"/>
    <origin rpy="0 0 0" xyz="0 0 0"/>
  </joint>
</robot>

